arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -902,6 +910,28 @@ Result IResearchDataStore::deleteDataStore() noexcept {
   return {};
 }
 
+void IResearchDataStore::setFailed() noexcept {
+  // should never be called on coordinators, only on DB servers and
+  // single servers
+  TRI_ASSERT(!ServerState::instance()->isCoordinator());
+
+  // this is expected to be set only once during the recovery phase or
+  // link loading phase.
+  if (!_failed.exchange(true)) {
```

Comment:
why exchange and why atomics? Is setFailed called from multiple threads?

Reply:
`setFailed()` can currently be called from `init()` and from the recovery callback.
I think in both cases there is just a single thread accessing the link, but that's speculation from my side.
hasFailed() will later be called for preparing snapshots and for link serialization.
If you are sure it's fine to change `_failed` to non-atomic boolean, I am happy to change it.